### ISSUE TITLE
ARCH-2 — The backbone: interaction + symmetric contraction on the backend interface, `BaseMACE` assembly, descriptor API

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1561]
**Depends on:** ARCH-1 (#1559), BKD-1 (#1560) · **Blocks:** ARCH-3 (#1562), BKD-3 (#1567)

**Context:** Today the message-passing backbone lives inside monolithic model classes: `MACE.forward` (`mace/modules/models.py:47`) chains embeddings, one of eight `*InteractionBlock` variants (`mace/modules/blocks.py:835,938,1041,1162,1286,1412` for the six agnostic ones, plus the two magnetic spin-orbit-coupled variants at `:1632,1785` behind the shared `MagneticInteractionBlock:1495`), and `EquivariantProductBasisBlock` (`mace/modules/blocks.py:440`, with the magmom-conditioned `EquivariantProductBasisWithSelfMagmomBlock` at `:517`) wrapping a `SymmetricContraction` built with `CodeGenMixin`/fx codegen (`mace/modules/symmetric_contraction.py:26`, `torch.fx.symbolic_trace` at `:140,179,184`). The backend choice leaks into the blocks (`hasattr(self, "conv_fusion")` at `mace/modules/blocks.py:918,1021,1140,1264,1393,1577`), and LAMMPS real-vs-ghost partitioning is smuggled in via `lammps_class`/`lammps_natoms` attributes inside `InteractionBlock.handle_lammps` (`mace/modules/blocks.py:786-811`), whose entry guard is `if lammps_class is None or first_layer or torch.jit.is_scripting()` (`:793`) and whose halo exchange is `LAMMPS_MP.apply` at `:801` and `:810`. Descriptors — the backbone's raison d'être — are an afterthought: the model returns a raw `node_feats` entry consumed by `MACECalculator.get_descriptors` (`mace/calculators/mace.py:953-1003`, and a second copy on the cueq calculator at `:1565-1544`), the eval CLI's `--return_descriptors`/`--descriptor_num_layers`/`--descriptor_aggregation_method`/`--descriptor_invariants_only` flags (`mace/cli/eval_configs.py:73-94`, consumed at `:316-348` and `:454-469`), and the FPS fine-tuning selection (`mace/cli/fine_tuning_select.py:240-267`).

This ticket builds the v1 backbone: the interaction block (tensor-product convolution + residual variants) and the symmetric contraction reimplemented **on the BKD-1 (#1560) kernel-backend interface** (no e3nn, no fx codegen, no TorchScript), assembled into a two-layer-architecture `MACEBackbone` — node features in, node features out, **no readouts, no gradient calls** — with a first-class descriptor-extraction API and an equivariance test suite. Legacy is read in place as the behavioural oracle; reproduce its numbers, never its structure.

**Interface & constraints:**

1. **Placement** (all new code, zero edits to frozen legacy `mace/`): `packages/mace-torch/…/nn/interaction.py`, `nn/product_basis.py`, `nn/symmetric_contraction.py`, `models/base.py`. Blocks are born **without `@compile_mode("script")`** — TorchScript is banned in `packages/`; compile-first with eager as the always-working reference. Descriptive names (`sender_node_features`, not `x_s`); paper symbols go in docstrings.

2. **`MACEBackbone` contract** — an equivariant message-passing backbone, i.e. a learned descriptor:

```python
class MACEBackbone(torch.nn.Module):
    """Node features in, node features out. No readouts. No gradient calls."""
    def forward(self, graph: dict[str, Tensor]) -> NodeFeatures: ...
```

- Assembly: node embedding + radial embedding + spherical harmonics come from ARCH-1 (#1559)'s ported blocks; then `num_interactions` × (interaction block → product-basis block). Per-layer node features are retained (they are the descriptors, see 6).
- The forward consumes the **flat-dict graph contract**. Keys the backbone reads (`F` = model dtype, `float64` default):

| key | shape | dtype | notes |
|---|---|---|---|
| `positions` | `[n_nodes, 3]` | F | Å |
| `node_attrs` | `[n_nodes, n_elements]` | F | one-hot vs the model's z-table |
| `edge_index` | `[2, n_edges]` | int64 | `[0]=sender, [1]=receiver` |
| `shifts` | `[n_edges, 3]` | F | Å; recomputed upstream under strain |
| `batch` | `[n_nodes]` | int64 | node → graph id, non-decreasing |
| `ptr` | `[n_graphs + 1]` | int64 | `num_graphs := ptr.numel() - 1` is the **only** sanctioned derivation (legacy `mace/modules/utils.py:811`, `mace/modules/models.py:786,1073,1349`, `mace/modules/extensions.py:1765`) |

- **Read-only forward:** the model never writes into the input dict. Schema validation (`validate_graph`) runs at boundaries, **never inside `forward`**.
- `edge_vectors`/`edge_lengths` are deliberately **not** input keys: the backbone computes them from `positions`/`edge_index`/`shifts`. Strained positions/shifts arrive as ordinary inputs — ARCH-4 (#1563)'s derivative engine composes *around* the backbone (the strain displacement is injected before edge vectors are computed; that seam design lives in ARCH-4 (#1563) and must not be foreclosed here).
- **Conditioned node inputs are a declared extension of this table, not an exception to it.** Develop feeds `magmom` (`[n_nodes, 3]`) into the backbone as a second node attribute stream (`mace/modules/extensions.py:1834` → `EquivariantProductBasisWithSelfMagmomBlock`), and it is a *grad leaf*. `MACEBackbone` must therefore accept additional declared node-input keys without the assembly special-casing any of them; the differentiation of such an input is ARCH-5 (#1633)'s, never the backbone's.
- **Padding-unaware:** a padded batch is a valid batch (inert padding — element-0 atoms, self-loop edges at length `2·r_max` where the polynomial cutoff is exactly 0). No masks, no real-count reads in the forward; outputs on real slots must be identical with and without padding.

3. **Every equivariant op goes through the BKD-1 (#1560) backend factories, resolved once at model build time** — nothing resolves in `forward`, no dtype/device/isinstance checks on the hot path, no direct tensor-product math in block code:

- **Interaction block** = `linear` (up-projection) → `channelwise_tp_conv` (node-level: returns `[n_nodes, …]`; radial-MLP weights are external, in canonical segment order; whether the scatter is fused is backend-internal — there is **no `conv_fusion` branch and no `.forward` monkeypatch anywhere**) → `linear` → density normalization by `avg_num_neighbors` (legacy `message = self.linear(message) / self.avg_num_neighbors`, `mace/modules/blocks.py:929,1033,1404`; the Density variants divide by a *learned* per-node density instead, which is why they are a separate variant and not a flag) → `fully_connected_tp` (skip_tp) against `node_attrs`. The radial MLP itself is a plain dense net, not a dispatched op.
- **Product-basis block** = `symmetric_contraction` op (weights internal, per-element, canonical flat `[Z, A, mul]` over the **reduced Clebsch-Gordan basis** pinned by CORE-4 (#1558)) + `linear`, with the residual self-connection added after the contraction. No fx tracing, no `CodeGenMixin`.
- **One activation layout per resolved chain; no layout transpose in `forward`** — layout reshapes live only at the backend boundary. Legacy pays exactly this cost in the hybrid path, where `layout_transpose_in`/`layout_transpose_out` are applied inside the patched forward (`mace/modules/wrapper_ops.py:231-232,240-241`).

4. **Which interaction variants:** `RealAgnosticInteractionBlock` (`mace/modules/blocks.py:835`, first-layer) and `RealAgnosticResidualInteractionBlock` (`mace/modules/blocks.py:938`, standard) are KEEP and in scope. The four REVIEW variants carry their inventory decisions: `RealAgnosticDensityInteractionBlock` (`mace/modules/blocks.py:1041`) / `RealAgnosticDensityResidualInteractionBlock` (`mace/modules/blocks.py:1162`) — KEEP iff a published foundation model uses them (then they need a golden); `RealAgnosticAttResidualInteractionBlock` (`mace/modules/blocks.py:1286`) — research variant; `RealAgnosticResidualNonLinearInteractionBlock` (`mace/modules/blocks.py:1412`) — new, owner unresolved. All six are reachable from the CLI through `interaction_classes` (`mace/modules/__init__.py:71-80`), which also registers the two magnetic variants — those are ARCH-5 (#1633)/MAG-1 (#1584)'s, and this ticket's composition must admit them rather than re-deriving them. A non-linear **first** interaction block is permitted in `MACE`-type models (legacy #1394, commit ff14d0b) — the composition must support it; the config flag itself lands in CFG-1 (#1574). The SO(3)-vs-O(3) mode (`--use_so3`, `arg_parser.py:246`) and `--use_agnostic_product` (`:252`) are REVIEW items resolved inside this ticket; neither is exercised today (`tests/backends/backend_parity.py:128-129` parametrizes `use_agnostic_product` and `use_last_readout_only` over `[False]` only).

5. **Ghost-atom seam designed in, not bolted on.** LAMMPS domain decomposition needs a halo exchange of node features **between interaction layers** (legacy applies it inside `handle_lammps`, `mace/modules/blocks.py:786-811`, driven by `lammps_class` set only by `mace/calculators/lammps_mliap_mace.py:281`, guarded by `torch.jit.is_scripting()` at `blocks.py:793` — that probe dies with TorchScript). Two develop facts bound the design: the exchange is `LAMMPS_MP.apply(node_feats, lammps_class)` and needs LAMMPS's `forward_exchange`, which exists only in the KOKKOS ML-IAP coupling, so `LAMMPS_MLIAP_MACE` refuses a multi-layer model up front (`mace/calculators/lammps_mliap_mace.py:187-205`) rather than dying inside layer two. In v1: ghost atoms are trailing node slots plus a `LocalityInfo` sidecar (`n_local`, `n_ghost`) — never new schema keys; the backbone slices `[:n_local]` where legacy's `truncate_ghosts` (`blocks.py:813-817`) does. **`LocalityInfo` is per-call data, not build-time configuration**: `MACEBackbone.forward(self, graph: dict[str, Tensor], locality: LocalityInfo | None = None)` takes it as an optional second argument, defaulting to `None` for the single-domain case, so the graph dict gains no keys and nothing is ever assigned onto the module between calls. What is fixed at build time is only the hook itself; and `MACEBackbone` exposes an **explicit between-layer hook** (a communication seam between compiled regions, not an op) whose presence is an explicit configuration — never inferred from compile mode. Default is a no-op with a bit-identical forward. The actual MPI exchange, the multi-layer refusal, and export wiring are DEP-2 (#1594).

6. **Descriptor-extraction API — MACEBackbone features ARE the descriptors.** A first-class API over the per-layer node features, with legacy-compatible semantics (behavioural spec: `mace/calculators/mace.py:953-1003` and `mace/cli/eval_configs.py:316-348,454-469`): `num_layers` (`-1` = all interaction layers), `invariants_only` (keep only `l=0` channels per layer, legacy `extract_invariant`, `mace/modules/utils.py:460`), and an aggregation method (`None` = per-atom descriptors; `"mean"` and `"per_element_mean"` are the two legacy reductions, `eval_configs.py:456-465`). Two legacy details are part of the contract, not accidents: the **last layer carries only invariant features**, so the per-layer width table is `[irreps_out.dim] * num_interactions` with the last entry replaced by `num_invariant_features` (`mace/calculators/mace.py:980-985`), and the truncation is `descriptor[:, :sum(per_layer_features[:num_layers])]`. Consumers to serve: eval-CLI descriptor output, the calculator's `get_descriptors` (shape `(num_atoms, invariant_features_up_to_num_layers)` for a single model, a list otherwise), and FPS-based fine-tuning subset selection. Legacy restricts `get_descriptors` to `model_type == "MACE"` (`mace/calculators/mace.py:964`); v1 serves any model with a backbone, and the restriction becomes a **recorded DROP** — it was a consequence of descriptors living on the calculator, not a semantic limit.

**Note on naming, settled:** this ticket used to call the backbone `MACEBackbone`, and `docs/reforge/target_layout.md` used the same name for the composed model, so the two could not both hold. The name is retired rather than assigned to either. The backbone is **`MACEBackbone`** in `nn/backbone.py`; the composed model that holds it beside ARCH-3 (#1562)'s output layer is **`MACEModel`** in `models/base.py`, and that is the class a new model subclasses, because the term a model like `PolarMACE` adds is an energy and an energy does not exist until the output layer has run. The layout document, the extension guides and the plan now use these two names. The fitness function that forced the question is fixed too: it keyed on a class whose name ends in `MACE`, which both caught a backbone that has no typed output and missed the composed model entirely, so `test_construction_via_modelconfig` had been passing on an empty set.

**Task:**
1. `nn/symmetric_contraction.py` + `nn/product_basis.py`: `EquivariantProductBasisBlock` on the `symmetric_contraction` backend op (reduced basis, canonical `[Z, A, mul]`), residual add, no layout transpose in forward.
2. `nn/interaction.py`: the KEEP interaction blocks (first-layer + residual) on the backend factories — fusion-agnostic forward calling node-level `channelwise_tp_conv`, `avg_num_neighbors` density normalization, `fully_connected_tp` skip connection; resolve the REVIEW variants (Density/Att/NonLinear, `--use_so3`, `--use_agnostic_product`) and record the outcomes.
3. `nn/backbone.py`: `MACEBackbone` assembling ARCH-1 (#1559) embeddings + interaction/product stacks from the flat-dict; read-only forward; edge vectors computed internally; per-layer feature retention; support for declared additional node-input streams; the explicit between-layer locality/halo hook (default no-op).
4. The descriptor-extraction API (`num_layers` / `invariants_only` / `None|mean|per_element_mean` aggregation) with legacy-parity semantics, including the invariant-only last layer.
5. Equivariance test suite: rotation (features transform by the Wigner-D matrices of their irreps; `l=0` channels invariant), parity/inversion, translation invariance, node-permutation equivariance — for each interaction variant in scope.
6. Backbone parity vs the frozen legacy `mace/` `node_feats` on the tiny anchors at the fp64 tolerance row, both interaction variants.

**Out of scope:** readouts, heads, scale-shift, E0s, observables (ARCH-3 (#1562)); the derivative engine and the strain seam (ARCH-4 (#1563)); conditioned-input differentiation and the magnetic product-basis variant (ARCH-5 (#1633)); accelerated backends and their parity (BKD-3 (#1567)); the compiled padded path + PrecisionConfig (BKD-2 (#1566)); collation/neighbor lists (DATA-3 (#1571)); the MLIAP export, the multi-layer refusal, and real MPI halo exchange (DEP-2 (#1594)); the fine-tuning selection CLI itself (FT-1 (#1580)).

**Acceptance criteria:**
- [ ] Backbone node features match the frozen legacy `node_feats` on the tiny anchors at the fp64 tolerance row, for both the first-layer and residual interaction variants.
- [ ] Equivariance suite passes at fp64: rotated inputs give Wigner-D-transformed features, invariant channels unchanged, translation-invariant, permutation-equivariant.
- [ ] The forward performs no backend resolution, no layout transpose, and never writes into the input graph dict (asserted by test).
- [ ] No `conv_fusion` attribute branch, no `.forward` monkeypatch, no `torch.jit.is_scripting()` probe, no `@compile_mode` anywhere in the new code.
- [ ] Descriptor API reproduces legacy `get_descriptors` outputs (all `invariants_only` × `num_layers` combinations, plus both aggregation methods) on the anchors, including the invariant-only last-layer width.
- [ ] Padded vs unpadded batches give identical features on real slots.
- [ ] With the locality profile configured, the backbone slices `[:n_local]` where legacy does; with it absent, the forward is bit-identical to the hook-free path.
- [ ] `torch.compile(fullgraph=True)` of `MACEBackbone` on the reference backend runs without graph breaks (eager remains the reference path).
- [ ] REVIEW decisions (Density/Att/NonLinear variants, `--use_so3`, `--use_agnostic_product`) are recorded in the PR with their rationale.

**Inventory gaps assigned here:**
- [ ] `--use_edge_irreps_first` (`mace/tools/arg_parser.py:285`) — the first-layer variant of `--edge_irreps` (`:279`); the base flag is pinned by the MACE-Polar golden but this one is not, and neither has a test.
- [ ] The non-linear first interaction block (#1394) is reachable from config and covered.
- [ ] `--use_so3` and `--use_agnostic_product` are untested today; whichever disposition they get, the KEEP ones gain cases here and the DROP ones a one-line justification.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/nn/test_interaction.py packages/mace-torch/tests/nn/test_product_basis.py -v
python -m pytest packages/mace-torch/tests/models/test_base_mace.py -v      # equivariance, read-only forward, padding, descriptors
python -m pytest tests/parity -m "not slow" -k backbone                     # node_feats vs frozen legacy anchors
```
**Review focus:** that every equivariant op reaches the backend through the BKD-1 (#1560) factories (nothing resolves in `forward`, no hand-rolled tensor products in block code); the equivariance suite's irrep coverage; descriptor-semantics parity with the legacy calculator; and that the between-layer halo hook is explicit configuration, never a compile-mode probe.

---
